diff --git a/cupy/cuda/jitify.pyx b/cupy/cuda/jitify.pyx
index b27e3d4ae27..41e3f6fe117 100644
--- a/cupy/cuda/jitify.pyx
+++ b/cupy/cuda/jitify.pyx
@@ -42,7 +42,7 @@ cdef extern from 'cupy_jitify.h' namespace "jitify::detail" nogil:
 # users.
 cdef extern from *:
     """
-    const int build_num = 2;
+    const int build_num = 3;
     """
     const int build_num
 
@@ -139,7 +139,10 @@ cdef str warmup_kernel = r"""cupy_jitify_exercise
 
 #include <cub/block/block_reduce.cuh>
 #include <cub/block/block_load.cuh>
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 700
+// not supported before CC 7.0
 #include <cuda/barrier>
+#endif
 #include <cooperative_groups.h>
 #include <cooperative_groups/memcpy_async.h>
 
