diff --git a/.travis.yml b/.travis.yml
index 95d199c16..b9e50ad12 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -68,7 +68,13 @@ jobs:
         # Install rasterio using pip to avoid conda solver errors
         - pip install rasterio
         - doit env_capture
+        - conda list --full-name cupy --json
       script:
+        - conda install dask>=2.0.0
+        # https://numba.pydata.org/numba-doc/latest/user/installing.html#installing-from-source
+        - pip install git+https://github.com/stuartarchibald/numba.git@fix/cuda_atomic_nanminmax
+        # - git clone git://github.com/numba/numba.git
+        # - cd numba && python setup.py install && cd ..
         - doit test_all
         - export NUMBA_DISABLE_JIT=1; doit test_unit_nojit
       after_success: codecov
@@ -83,6 +89,10 @@ jobs:
     - <<: *default
       env: PYTHON_VERSION=3.7
       script:
+        # https://numba.pydata.org/numba-doc/latest/user/installing.html#installing-from-source
+        - pip install git+https://github.com/stuartarchibald/numba.git@fix/cuda_atomic_nanminmax
+        # - git clone git://github.com/numba/numba.git
+        # - cd numba && python setup.py install && cd ..
         - doit test_all
         - export NUMBA_DISABLE_JIT=1; doit test_unit_nojit
 
diff --git a/datashader/reductions.py b/datashader/reductions.py
index 08ac3a9a4..c2b11c79c 100644
--- a/datashader/reductions.py
+++ b/datashader/reductions.py
@@ -8,6 +8,8 @@
 
 from datashader.glyphs.glyph import isnull
 from numba import cuda as nb_cuda
+#from datashader.transfer_functions._cuda_utils import (cuda_atomic_nanmin,
+#                                                       cuda_atomic_nanmax)
 
 try:
     import cudf
@@ -479,7 +481,7 @@ def _append(x, y, agg, field):
     @staticmethod
     @ngjit
     def _append_cuda(x, y, agg, field):
-        nb_cuda.atomic.min(agg, (y, x), field)
+        nb_cuda.atomic.nanmax(agg, (y, x), field)
 
     @staticmethod
     def _combine(aggs):
@@ -506,7 +508,7 @@ def _append(x, y, agg, field):
     @staticmethod
     @ngjit
     def _append_cuda(x, y, agg, field):
-        nb_cuda.atomic.max(agg, (y, x), field)
+        nb_cuda.atomic.nanmax(agg, (y, x), field)
 
     @staticmethod
     def _combine(aggs):
diff --git a/datashader/transfer_functions/_cuda_utils.py b/datashader/transfer_functions/_cuda_utils.py
index 523f0ebe5..4e2378db5 100644
--- a/datashader/transfer_functions/_cuda_utils.py
+++ b/datashader/transfer_functions/_cuda_utils.py
@@ -1,6 +1,8 @@
 from __future__ import division
+from distutils.version import LooseVersion
 from math import ceil, isnan, nan
 from numba import cuda
+import numba
 import cupy
 import numpy as np
 
@@ -70,13 +72,35 @@ def masked_clip_2d(data, mask, lower, upper):
     masked_clip_2d_kernel[cuda_args(data.shape)](data, mask, lower, upper)
 
 
+# Behaviour of numba.cuda.atomic.max/min changed in 0.50 so as to behave as per
+# np.nanmax/np.nanmin
+if LooseVersion(numba.__version__) >= LooseVersion("0.51.0"):
+    @cuda.jit
+    def cuda_atomic_nanmin(ary, idx, val):
+        return cuda.atomic.nanmin(ary, idx, val)
+    @cuda.jit
+    def cuda_atomic_nanmax(ary, idx, val):
+        return cuda.atomic.nanmax(ary, idx, val)
+elif LooseVersion(numba.__version__) <= LooseVersion("0.49.0"):
+    @cuda.jit
+    def cuda_atomic_nanmin(ary, idx, val):
+        return cuda.atomic.min(ary, idx, val)
+
+    @cuda.jit
+    def cuda_atomic_nanmax(ary, idx, val):
+        return cuda.atomic.max(ary, idx, val)
+else:
+    # WHAT DO YOU WANT TO DO HERE?
+    pass
+
+
 @cuda.jit
 def masked_clip_2d_kernel(data, mask, lower, upper):
     i, j = cuda.grid(2)
     maxi, maxj = data.shape
     if i >= 0 and i < maxi and j >= 0 and j < maxj and not mask[i, j]:
-        cuda.atomic.max(data, (i, j), lower)
-        cuda.atomic.min(data, (i, j), upper)
+        cuda_atomic_nanmax(data, (i, j), lower)
+        cuda_atomic_nanmin(data, (i, j), upper)
 
 
 # interp
diff --git a/setup.py b/setup.py
index 830c9a678..d4cd3fdda 100644
--- a/setup.py
+++ b/setup.py
@@ -15,7 +15,9 @@
     'dask[complete] >=0.18.0',
     'toolz >=0.7.4',  # ? for some dask issue (dasks does only >=0.7.3)
     'datashape >=0.5.1',
-    'numba >=0.37.0,<0.49',
+    # 'numba >=0.37.0,<0.49',
+    # 'numba @ git+https://github.com/stuartarchibald/numba.git@fix/cuda_atomic_nanminmax',
+    # 'numba >=0.37.0', # maybe this will install whatever cupy we need, then we can overwrite it with numba from git without llvmlite conflicting because this won't install unremovable llvmlite 0.31.0?
     'numpy >=1.7',
     'pandas >=0.24.1',
     'pillow >=3.1.1',
