diff --git a/aten/src/ATen/DLConvertor.cpp b/aten/src/ATen/DLConvertor.cpp
index c4b4d9bc086b5..85caa564c08b1 100644
--- a/aten/src/ATen/DLConvertor.cpp
+++ b/aten/src/ATen/DLConvertor.cpp
@@ -70,6 +70,9 @@ DLDataType getDLDataType(const Tensor& t) {
     case ScalarType::Float8_e8m0fnu:
       TORCH_CHECK(false, "float8 types are not supported by dlpack");
       break;
+    case ScalarType::Float4_e2m1fn_x2:
+      TORCH_CHECK(false, "float4 types are not supported by dlpack");
+      break;
     case ScalarType::QInt8:
     case ScalarType::QUInt8:
     case ScalarType::QInt32:
diff --git a/c10/core/ScalarType.cpp b/c10/core/ScalarType.cpp
index ae44dd556beb1..b14f209359cb5 100644
--- a/c10/core/ScalarType.cpp
+++ b/c10/core/ScalarType.cpp
@@ -224,6 +224,8 @@ std::pair<std::string, std::string> getDtypeNames(c10::ScalarType scalarType) {
       return std::make_pair("float8_e4m3fnuz", "");
     case c10::ScalarType::Float8_e8m0fnu:
       return std::make_pair("float8_e8m0fnu", "");
+    case c10::ScalarType::Float4_e2m1fn_x2:
+      return std::make_pair("float4_e2m1fn_x2", "");
     default:
       throw std::runtime_error("Unimplemented scalar type");
   }
diff --git a/c10/core/ScalarType.h b/c10/core/ScalarType.h
index 1b4fa9aae7507..51579d3fe64c5 100644
--- a/c10/core/ScalarType.h
+++ b/c10/core/ScalarType.h
@@ -3,6 +3,7 @@
 #include <c10/util/BFloat16.h>
 #include <c10/util/Deprecated.h>
 #include <c10/util/Exception.h>
+#include <c10/util/Float4_e2m1fn_x2.h>
 #include <c10/util/Float8_e4m3fn.h>
 #include <c10/util/Float8_e4m3fnuz.h>
 #include <c10/util/Float8_e5m2.h>
@@ -104,7 +105,8 @@ struct dummy_int1_7_t {};
   _(c10::dummy_int1_7_t<5>, Int5) /* 41 */               \
   _(c10::dummy_int1_7_t<6>, Int6) /* 42 */               \
   _(c10::dummy_int1_7_t<7>, Int7) /* 43 */               \
-  _(c10::Float8_e8m0fnu, Float8_e8m0fnu) /* 44 */
+  _(c10::Float8_e8m0fnu, Float8_e8m0fnu) /* 44 */        \
+  _(c10::Float4_e2m1fn_x2, Float4_e2m1fn_x2) /* 45 */
 
 // If you want to support ComplexHalf for real, add ComplexHalf
 // into this macro (and change the name).  But beware: convert()
@@ -387,7 +389,8 @@ inline bool isFloat8Type(ScalarType t) {
 }
 
 inline bool isReducedFloatingType(ScalarType t) {
-  return t == ScalarType::Half || t == ScalarType::BFloat16 || isFloat8Type(t);
+  return t == ScalarType::Half || t == ScalarType::BFloat16 ||
+      isFloat8Type(t) || t == ScalarType::Float4_e2m1fn_x2;
 }
 
 inline bool isFloatingType(ScalarType t) {
@@ -498,6 +501,7 @@ inline bool isSignedType(ScalarType t) {
     case ScalarType::Int5:
     case ScalarType::Int6:
     case ScalarType::Int7:
+    case ScalarType::Float4_e2m1fn_x2:
       return true;
     case ScalarType::UInt1:
     case ScalarType::UInt2:
diff --git a/c10/util/Float4_e2m1fn_x2.h b/c10/util/Float4_e2m1fn_x2.h
new file mode 100644
index 0000000000000..4ba1d77721178
--- /dev/null
+++ b/c10/util/Float4_e2m1fn_x2.h
@@ -0,0 +1,24 @@
+#pragma once
+#include <cstdint>
+
+#include <c10/macros/Macros.h>
+
+/// Defines the Float4_e2m1fn_x2 type (4-bit floating-point, two elements packed
+/// into one byte) Binary configuration :
+///
+///   0123:4567
+///   seem:seem
+///
+/// This is the FP4 dtype from the OCP MX format spec
+/// (https://www.opencompute.org/documents/ocp-microscaling-formats-mx-v1-0-spec-final-pdf,
+/// Section 5.3.3)
+
+namespace c10 {
+
+struct alignas(1) Float4_e2m1fn_x2 {
+  uint8_t val_;
+  Float4_e2m1fn_x2() = default;
+  C10_HOST_DEVICE explicit Float4_e2m1fn_x2(uint8_t val) : val_(val) {}
+};
+
+} // namespace c10
diff --git a/test/quantization/core/experimental/test_float8.py b/test/quantization/core/experimental/test_floatx.py
similarity index 96%
rename from test/quantization/core/experimental/test_float8.py
rename to test/quantization/core/experimental/test_floatx.py
index 1c4956d551a44..b87e6ab67c601 100644
--- a/test/quantization/core/experimental/test_float8.py
+++ b/test/quantization/core/experimental/test_floatx.py
@@ -392,7 +392,23 @@ def test_save_load(self, dtype, device):
             torch.testing.assert_close(x1, x1_save_load, atol=0, rtol=0)
 
 
+class TestFloat4Dtype(TestCase):
+    def test_float4_e2m1fn_x2(self, device):
+        # can create a tensor of dtype float4
+        x1 = torch.empty(4, 4, device=device, dtype=torch.float4_e2m1fn_x2)
+
+        # can create a string (so printing will work)
+        str(x1)
+
+        # can view float4_e2m1fn_x2 as uint8
+        x2 = x1.view(torch.uint8)
+
+        # can view uint8 as float4_e2m1fn_x2
+        x2.view(torch.float4_e2m1fn_x2)
+
+
 instantiate_device_type_tests(TestFloat8Dtype, globals())
+instantiate_device_type_tests(TestFloat4Dtype, globals())
 
 
 class TestFloat8DtypeCPUOnly(TestCase):
diff --git a/test/test_quantization.py b/test/test_quantization.py
index 61a8e310c7afc..70bde60406aae 100644
--- a/test/test_quantization.py
+++ b/test/test_quantization.py
@@ -163,15 +163,15 @@
 except ImportError as e:
     logging.warning(e)
 try:
-    from quantization.core.experimental.test_float8 import TestFloat8DtypeCPU  # noqa: F401
+    from quantization.core.experimental.test_floatx import TestFloat8DtypeCPU  # noqa: F401
 except ImportError as e:
     logging.warning(e)
 try:
-    from quantization.core.experimental.test_float8 import TestFloat8DtypeCUDA  # noqa: F401
+    from quantization.core.experimental.test_floatx import TestFloat8DtypeCUDA  # noqa: F401
 except ImportError as e:
     logging.warning(e)
 try:
-    from quantization.core.experimental.test_float8 import TestFloat8DtypeCPUOnlyCPU  # noqa: F401
+    from quantization.core.experimental.test_floatx import TestFloat8DtypeCPUOnlyCPU  # noqa: F401
 except ImportError as e:
     logging.warning(e)
 
diff --git a/torch/_tensor_str.py b/torch/_tensor_str.py
index 180d8e9001ff2..068048a7ec7ff 100644
--- a/torch/_tensor_str.py
+++ b/torch/_tensor_str.py
@@ -142,6 +142,14 @@ def __init__(self, tensor):
                 self.max_width = max(self.max_width, len(value_str))
 
         else:
+            if tensor.dtype == torch.float4_e2m1fn_x2:  # type: ignore[attr-defined]
+                # torch.float4_e2m1fn_x2 is special and does not support the casts necessary
+                # to print it, we choose to display the uint8 representation here for
+                # convenience of being able to print a tensor.
+                # TODO(#146647): extend this to other dtypes without casts defined, such
+                # as the bits, uint1..7 and int1..7 dtypes.
+                tensor_view = tensor_view.view(torch.uint8)
+
             nonzero_finite_vals = torch.masked_select(
                 tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0)
             )
@@ -258,6 +266,14 @@ def _val_formatter(val, formatter1=formatter1, formatter2=formatter2):
         else:
             return formatter1.format(val)
 
+    if self.dtype == torch.float4_e2m1fn_x2:  # type: ignore[attr-defined]
+        # torch.float4_e2m1fn_x2 is special and does not support the casts necessary
+        # to print it, we choose to display the uint8 representation here for
+        # convenience of being able to print a tensor.
+        # TODO(#146647): extend this to other dtypes without casts defined, such
+        # as the bits, uint1..7 and int1..7 dtypes.
+        self = self.view(torch.uint8)
+
     if summarize and not PRINT_OPTS.edgeitems:
         # Deal with edge case that negative zero is zero
         data = ["..."]
