diff --git a/CHANGELOG.md b/CHANGELOG.md
index c660d83500b..0da9c611ff9 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -144,6 +144,7 @@
 - PR #4339 Port libcudf strings `wrap` api to cython/python
 - PR #4311 Port nvstrings String Manipulations functions to cuDF Python/Cython
 - PR #4373 Port nvstrings Regular Expressions functions to cuDF Python/Cython
+- PR #4377 Support loading avro files that contain nested arrays
 
 ## Bug Fixes
 
diff --git a/cpp/src/io/avro/avro.cpp b/cpp/src/io/avro/avro.cpp
index 953883ed6b8..e035987bb30 100644
--- a/cpp/src/io/avro/avro.cpp
+++ b/cpp/src/io/avro/avro.cpp
@@ -153,7 +153,8 @@ bool container::parse(file_metadata *md, size_t max_num_rows, size_t first_row)
                             } while (skip != 0);
                         }
                     }
-                    if (parent_idx != 0 || col.name.length() == 0) // Ignore the root entry
+                    // Ignore the root or array entries
+                    if ((parent_idx != 0 && md->schema[parent_idx].kind != type_array) || col.name.length() == 0)
                     {
                         if (col.name.length() > 0)
                         {
@@ -164,11 +165,9 @@ bool container::parse(file_metadata *md, size_t max_num_rows, size_t first_row)
                     parent_idx = md->schema[parent_idx].parent_idx;
                 }
             }
-            md->columns.push_back(col);
+            md->columns.emplace_back(std::move(col));
         }
     }
-
-
     return true;
 }
 
@@ -192,6 +191,7 @@ enum {
     attrtype_name,
     attrtype_fields,
     attrtype_symbols,
+    attrtype_items,
 };
 
 /**
@@ -208,12 +208,13 @@ bool schema_parser::parse(std::vector<schema_entry> &schema, const std::string &
     int depth = 0, parent_idx = -1, entry_idx = -1;
     json_state_e state = state_attrname;
     std::string str;
-    std::unordered_map<std::string, type_kind_e> typenames = {
+    const std::unordered_map<std::string, type_kind_e> typenames = {
                                             {"null", type_null},    {"boolean", type_boolean},  {"int", type_int},      {"long", type_long},
                                             {"float", type_float},  {"double", type_double},    {"bytes", type_bytes},  {"string", type_string},
-                                            {"record", type_record},{"enum", type_enum}};
-    std::unordered_map<std::string, int> attrnames = {
-                                            {"type", attrtype_type}, {"name", attrtype_name}, {"fields", attrtype_fields}, {"symbols", attrtype_symbols}};
+                                            {"record", type_record},{"enum", type_enum},        {"array", type_array}};
+    const std::unordered_map<std::string, int> attrnames = {
+                                            {"type", attrtype_type}, {"name", attrtype_name}, {"fields", attrtype_fields},
+                                            {"symbols", attrtype_symbols}, {"items", attrtype_items}};
     int cur_attr = attrtype_none;
     m_base = json_str.c_str();
     m_cur = m_base;
@@ -259,7 +260,7 @@ bool schema_parser::parse(std::vector<schema_entry> &schema, const std::string &
                 {
                     if (entry_idx < 0)
                         return false;
-                    schema[entry_idx].name = str;
+                    schema[entry_idx].name = std::move(str);
                 }
                 if (state == state_attrvalue_last)
                 {
@@ -272,7 +273,7 @@ bool schema_parser::parse(std::vector<schema_entry> &schema, const std::string &
             {
                 if (entry_idx < 0)
                     return false;
-                schema[entry_idx].symbols.emplace_back(str);
+                schema[entry_idx].symbols.emplace_back(std::move(str));
             }
             break;
         case ':':
@@ -303,6 +304,14 @@ bool schema_parser::parse(std::vector<schema_entry> &schema, const std::string &
                 cur_attr = attrtype_none;
                 state = state_attrname;
             }
+            else if (state == state_attrvalue && cur_attr == attrtype_items && entry_idx >= 0)
+            {
+                // Treat array as a one-field record
+                parent_idx = entry_idx;
+                entry_idx = -1;
+                cur_attr = attrtype_none;
+                state = state_attrname;
+            }
             if (depth >= MAX_SCHEMA_DEPTH || state != state_attrname)
             {
                 return false;
diff --git a/cpp/src/io/avro/avro_common.h b/cpp/src/io/avro/avro_common.h
index 1c7de8c0858..7cbeb461ade 100644
--- a/cpp/src/io/avro/avro_common.h
+++ b/cpp/src/io/avro/avro_common.h
@@ -53,6 +53,7 @@ enum type_kind_e
     type_enum,
     type_record,
     type_union,
+    type_array,
 };
 
 
diff --git a/cpp/src/io/avro/avro_gpu.cu b/cpp/src/io/avro/avro_gpu.cu
index c7241742733..3f13d849e39 100644
--- a/cpp/src/io/avro/avro_gpu.cu
+++ b/cpp/src/io/avro/avro_gpu.cu
@@ -24,6 +24,212 @@ namespace gpu {
 #define NWARPS                  16
 #define MAX_SHARED_SCHEMA_LEN   1000
 
+/*
+ * Avro varint encoding - see
+ * https://avro.apache.org/docs/1.2.0/spec.html#binary_encoding
+ */
+static inline int64_t __device__ avro_decode_zigzag_varint(const uint8_t *&cur, const uint8_t *end) {
+  uint64_t u = 0;
+  if (cur < end) {
+    u = *cur++;
+    if (u > 0x7f) {
+      uint64_t scale = 128;
+      u &= 0x7f;
+      while (cur < end) {
+        uint32_t c = *cur++;
+        u += (c & 0x7f) * scale;
+        scale <<= 7;
+        if (c < 0x80)
+          break;
+      }
+    }
+  }
+  return (int64_t)((u >> 1u) ^ -(int64_t)(u & 1));
+}
+
+
+/**
+ * @brief Decode a row of values given an avro schema
+ *
+ * @param[in] schema Schema description
+ * @param[in] schema_g Global schema in device mem
+ * @param[in] schema_len Number of schema entries
+ * @param[in] row Current row
+ * @param[in] max_rows Total number of rows
+ * @param[in] cur Current input data pointer
+ * @param[in] end End of input data
+ * @param[in] global_Dictionary Global dictionary entries
+ *
+ * @return data pointer at the end of the row (start of next row)
+ *
+ **/
+static const uint8_t * __device__ avro_decode_row(
+    const schemadesc_s *schema, schemadesc_s *schema_g, uint32_t schema_len,
+    size_t row, size_t max_rows,
+    const uint8_t *cur, const uint8_t *end,
+    const nvstrdesc_s *global_dictionary, uint32_t num_dictionary_entries)
+{
+  uint32_t array_start = 0, array_repeat_count = 0;
+  int array_children = 0;
+  for (uint32_t i = 0; i < schema_len; ) {
+    uint32_t kind = schema[i].kind;
+    int skip = 0;
+    uint8_t *dataptr;
+    if (kind == type_union) {
+      int skip_after;
+      if (cur >= end)
+        break;
+      skip = (*cur++) >> 1;  // NOTE: Assumes 1-byte union member
+      skip_after = schema[i].count - skip - 1;
+      ++i;
+      while (skip > 0 && i < schema_len) {
+        if (schema[i].kind >= type_record) {
+          skip += schema[i].count;
+        }
+        ++i;
+        --skip;
+      }
+      if (i >= schema_len || skip_after < 0)
+        break;
+      kind = schema[i].kind;
+      skip = skip_after;
+    }
+    dataptr = reinterpret_cast<uint8_t *>(schema[i].dataptr);
+    switch (kind) {
+    case type_null:
+      if (dataptr && row < max_rows) {
+        atomicAnd(reinterpret_cast<uint32_t *>(dataptr) + (row >> 5), ~(1 << (row & 0x1f)));
+        atomicAdd(&schema_g[i].count, 1);
+      }
+      break;
+
+    case type_int:
+    case type_long:
+    case type_bytes:
+    case type_string:
+    case type_enum:
+      {
+        int64_t v = avro_decode_zigzag_varint(cur, end);
+        if (kind == type_int) {
+          if (dataptr && row < max_rows) {
+            reinterpret_cast<int32_t *>(dataptr)[row] = static_cast<int32_t>(v);
+          }
+        }
+        else if (kind == type_long) {
+          if (dataptr && row < max_rows) {
+            reinterpret_cast<int64_t *>(dataptr)[row] = v;
+          }
+        }
+        else { // string or enum
+          size_t count = 0;
+          const char *ptr = 0;
+          if (kind == type_enum) { // dictionary
+            size_t idx = schema[i].count + v;
+            if (idx < num_dictionary_entries) {
+              ptr = global_dictionary[idx].ptr;
+              count = global_dictionary[idx].count;
+            }
+          }
+          else if (v >= 0 && cur + v <= end) { // string
+            ptr = reinterpret_cast<const char *>(cur);
+            count = (size_t)v;
+            cur += count;
+          }
+          if (dataptr && row < max_rows) {
+            reinterpret_cast<nvstrdesc_s *>(dataptr)[row].ptr = ptr;
+            reinterpret_cast<nvstrdesc_s *>(dataptr)[row].count = count;
+          }
+        }
+      }
+      break;
+
+    case type_float:
+      if (dataptr && row < max_rows) {
+        uint32_t v;
+        if (cur + 3 < end) {
+          v = unaligned_load32(cur);
+          cur += 4;
+        }
+        else {
+          v = 0;
+        }
+        reinterpret_cast<uint32_t *>(dataptr)[row] = v;
+      }
+      else {
+        cur += 4;
+      }
+      break;
+
+    case type_double:
+      if (dataptr && row < max_rows) {
+        uint64_t v;
+        if (cur + 7 < end) {
+          v = unaligned_load64(cur);
+          cur += 8;
+        }
+        else {
+          v = 0;
+        }
+        reinterpret_cast<uint64_t *>(dataptr)[row] = v;
+      }
+      else {
+        cur += 8;
+      }
+      break;
+
+    case type_boolean:
+      if (dataptr && row < max_rows) {
+        uint8_t v = (cur < end) ? *cur : 0;
+        reinterpret_cast<uint8_t *>(dataptr)[row] = (v) ? 1 : 0;
+      }
+      cur++;
+      break;
+
+    case type_array:
+      {
+        int32_t array_block_count = avro_decode_zigzag_varint(cur, end);
+        if (array_block_count < 0) {
+          avro_decode_zigzag_varint(cur, end); // block size in bytes, ignored
+          array_block_count = -array_block_count;
+        }
+        array_start = i;
+        array_repeat_count = array_block_count;
+        array_children = 1;
+        if (array_repeat_count == 0) {
+          skip += schema[i].count; // Should always be 1
+        }
+      }
+      break;
+    }
+    if (array_repeat_count != 0) {
+      array_children--;
+      if (schema[i].kind >= type_record) {
+        array_children += schema[i].count;
+      }
+    }
+    i++;
+    while (skip > 0 && i < schema_len) {
+      if (schema[i].kind >= type_record) {
+        skip += schema[i].count;
+      }
+      ++i;
+      --skip;
+    }
+    // If within an array, check if we reached the last item
+    if (array_repeat_count != 0 && array_children <= 0 && cur < end) {
+      if (!--array_repeat_count) {
+        i = array_start; // Restart at the array parent
+      }
+      else {
+        i = array_start + 1; // Restart after the array parent
+        array_children = schema[array_start].count;
+      }
+    }
+  }
+  return cur;
+}
+
+
 /**
  * @brief Decode column data
  *
@@ -100,176 +306,10 @@ gpuDecodeAvroColumnData(block_desc_s *blocks, schemadesc_s *schema_g, nvstrdesc_
         }
         if (threadIdx.x < nrows)
         {
-            for (uint32_t i = 0; i < schema_len; )
-            {
-                uint32_t kind = schema[i].kind;
-                int skip = 0;
-                uint8_t *dataptr;
-                if (kind == type_union)
-                {
-                    int skip_after;
-                    if (cur >= end)
-                        break;
-                    skip = (*cur++) >> 1;  // NOTE: Assumes 1-byte union member
-                    skip_after = schema[i].count - skip - 1;
-                    ++i;
-                    while (skip > 0 && i < schema_len)
-                    {
-                        if (schema[i].kind >= type_record)
-                        {
-                            skip += schema[i].count;
-                        }
-                        ++i;
-                        --skip;
-                    }
-                    if (i >= schema_len || skip_after < 0)
-                        break;
-                    kind = schema[i].kind;
-                    skip = skip_after;
-                }
-                dataptr = reinterpret_cast<uint8_t *>(schema[i].dataptr);
-                if (dataptr)
-                {
-                    size_t row = cur_row - first_row + threadIdx.x;
-                    switch (kind)
-                    {
-                    case type_null:
-                        if (row < max_rows)
-                        {
-                            atomicAnd(reinterpret_cast<uint32_t *>(dataptr) + (row >> 5), ~(1 << (row & 0x1f)));
-                            atomicAdd(&schema_g[i].count, 1);
-                        }
-                        break;
-
-                    case type_int:
-                    case type_long:
-                    case type_bytes:
-                    case type_string:
-                    case type_enum:
-                        {
-                            uint64_t u = 0;
-                            int64_t v;
-                            if (cur < end)
-                            {
-                                u = *cur++;
-                                if (u > 0x7f)
-                                {
-                                    uint64_t scale = 128;
-                                    u &= 0x7f;
-                                    while (cur < end)
-                                    {
-                                        uint32_t c = *cur++;
-                                        u += (c & 0x7f) * scale;
-                                        scale <<= 7;
-                                        if (c < 0x80)
-                                            break;
-                                    }
-                                }
-                            }
-                            v = (int64_t)((u >> 1u) ^ -(int64_t)(u & 1));
-                            if (kind == type_int)
-                            {
-                                if (row < max_rows)
-                                {
-                                    reinterpret_cast<int32_t *>(dataptr)[row] = static_cast<int32_t>(v);
-                                }
-                            }
-                            else if (kind == type_long)
-                            {
-                                if (row < max_rows)
-                                {
-                                    reinterpret_cast<int64_t *>(dataptr)[row] = v;
-                                }
-                            }
-                            else // string or enum
-                            {
-                                size_t count = 0;
-                                const char *ptr = 0;
-                                if (kind == type_enum) // dictionary
-                                {
-                                    size_t idx = schema[i].count + v;
-                                    if (idx < num_dictionary_entries)
-                                    {
-                                        ptr = global_dictionary[idx].ptr;
-                                        count = global_dictionary[idx].count;
-                                    }
-                                }
-                                else if (v > 0 && cur + v <= end) // string
-                                {
-                                    ptr = reinterpret_cast<const char *>(cur);
-                                    count = (size_t)v;
-                                    cur += count;
-                                }
-                                if (row < max_rows)
-                                {
-                                    reinterpret_cast<nvstrdesc_s *>(dataptr)[row].ptr = ptr;
-                                    reinterpret_cast<nvstrdesc_s *>(dataptr)[row].count = count;
-                                }
-                            }
-                        }
-                        break;
-
-                    case type_float:
-                        {
-                            uint32_t v;
-                            if (cur + 3 < end)
-                            {
-                                v = (cur[3] << 24) | (cur[2] << 16) | (cur[1] << 8) | cur[0];
-                                cur += 4;
-                            }
-                            else
-                            {
-                                v = 0;
-                            }
-                            if (row < max_rows)
-                            {
-                                reinterpret_cast<uint32_t *>(dataptr)[row] = v;
-                            }
-                        }
-                        break;
-
-                    case type_double:
-                        {
-                            uint2 v;
-                            if (cur + 7 < end)
-                            {
-                                v.x = (cur[3] << 24) | (cur[2] << 16) | (cur[1] << 8) | cur[0];
-                                v.y = (cur[7] << 24) | (cur[6] << 16) | (cur[5] << 8) | cur[4];
-                                cur += 8;
-                            }
-                            else
-                            {
-                                v.x = v.y = 0;
-                            }
-                            if (row < max_rows)
-                            {
-                                reinterpret_cast<uint2 *>(dataptr)[row] = v;
-                            }
-                        }
-                        break;
-
-                    case type_boolean:
-                        {
-                            uint8_t v = (cur < end) ? *cur++ : 0;
-                            if (row < max_rows)
-                            {
-                                reinterpret_cast<uint8_t *>(dataptr)[row] = (v) ? 1 : 0;
-                            }
-                        }
-                        break;
-                    }
-                }
-                i++;
-                while (skip > 0 && i < schema_len)
-                {
-                    if (schema[i].kind >= type_record)
-                    {
-                        skip += schema[i].count;
-                    }
-                    ++i;
-                    --skip;
-                }
-            }
+            cur = avro_decode_row(schema, schema_g, schema_len,
+                                  cur_row - first_row + threadIdx.x,
+                                  max_rows, cur, end,
+                                  global_dictionary, num_dictionary_entries);
         }
         if (nrows <= 1)
         {
diff --git a/cpp/src/io/avro/reader_impl.cu b/cpp/src/io/avro/reader_impl.cu
index b67b44c0d8f..dfa00903d83 100644
--- a/cpp/src/io/avro/reader_impl.cu
+++ b/cpp/src/io/avro/reader_impl.cu
@@ -121,9 +121,19 @@ class metadata : public file_metadata {
       }
     } else {
       for (int i = 0; i < num_avro_columns; ++i) {
-        auto col_type = to_type_id(&schema[columns[i].schema_data_idx]);
-        CUDF_EXPECTS(col_type != type_id::EMPTY, "Unsupported data type");
-        selection.emplace_back(i, columns[i].name);
+        // Exclude array columns (unsupported)
+        bool column_in_array = false;
+        for (int parent_idx = schema[columns[i].schema_data_idx].parent_idx; parent_idx > 0; parent_idx = schema[parent_idx].parent_idx) {
+          if (schema[parent_idx].kind == avro::type_array) {
+            column_in_array = true;
+            break;
+          }
+        }
+        if (!column_in_array) {
+          auto col_type = to_type_id(&schema[columns[i].schema_data_idx]);
+          CUDF_EXPECTS(col_type != type_id::EMPTY, "Unsupported data type");
+          selection.emplace_back(i, columns[i].name);
+        }
       }
     }
     CUDF_EXPECTS(selection.size() > 0, "Filtered out all columns");
@@ -260,11 +270,12 @@ void reader::impl::decode_data(
   for (size_t i = 0; i < _metadata->schema.size(); i++) {
     type_kind_e kind = _metadata->schema[i].kind;
     if (skip_field_cnt != 0) {
-      // Exclude union members from min_row_data_size
+      // Exclude union and array members from min_row_data_size
       skip_field_cnt += _metadata->schema[i].num_children - 1;
     } else {
       switch (kind) {
         case type_union:
+        case type_array:
           skip_field_cnt = _metadata->schema[i].num_children;
           // fall through
         case type_boolean:
diff --git a/cpp/src/io/utilities/block_utils.cuh b/cpp/src/io/utilities/block_utils.cuh
index d242d9c8233..52fc1cb41fa 100644
--- a/cpp/src/io/utilities/block_utils.cuh
+++ b/cpp/src/io/utilities/block_utils.cuh
@@ -66,6 +66,25 @@ inline __device__ double Int128ToDouble_rn(uint64_t lo, int64_t hi)
     return sign * __fma_rn(__ll2double_rn(hi), 4294967296.0 * 4294967296.0, __ull2double_rn(lo));
 }
 
+inline __device__ uint32_t unaligned_load32(const uint8_t *p) {
+  uint32_t ofs = 3 & reinterpret_cast<uintptr_t>(p);
+  const uint32_t *p32 = reinterpret_cast<const uint32_t *>(p - ofs);
+  uint32_t v = p32[0];
+  return (ofs) ? __funnelshift_r(v, p32[1], ofs * 8) : v;
+}
+
+inline __device__ uint64_t unaligned_load64(const uint8_t *p) {
+  uint32_t ofs = 3 & reinterpret_cast<uintptr_t>(p);
+  const uint32_t *p32 = reinterpret_cast<const uint32_t *>(p - ofs);
+  uint32_t v0 = p32[0];
+  uint32_t v1 = p32[1];
+  if (ofs) {
+    v0 = __funnelshift_r(v0, v1, ofs * 8);
+    v1 = __funnelshift_r(v1, p32[2], ofs * 8);
+  }
+  return (((uint64_t)v1) << 32) | v0;
+}
+
 
 template<unsigned int nthreads, bool sync_before_store>
 inline __device__ void memcpy_block(void *dstv, const void *srcv, uint32_t len, uint32_t t)
